From 8d348e26c53c8a5267f1fe58e24f0649e845523f Mon Sep 17 00:00:00 2001
From: Harshitha Prem <quic_hprem@quicinc.com>
Date: Wed, 19 Oct 2022 23:29:38 +0530
Subject: [PATCH 6/7] nl80211: add support for background cac stop

Added changes to support aborting ongoing background
cac via new netlink command
NL80211_CMD_STOP_BGRADAR_DETECT.

Signed-off-by: Harshitha Prem <quic_hprem@quicinc.com>
Signed-off-by: Manish Dharanenthiran <quic_mdharane@quicinc.com>
---
 include/uapi/linux/nl80211.h |  2 ++
 net/wireless/nl80211.c       | 25 +++++++++++++++++++++++++
 2 files changed, 27 insertions(+)

diff --git a/include/uapi/linux/nl80211.h b/include/uapi/linux/nl80211.h
index d21c930..ee082c4 100644
--- a/include/uapi/linux/nl80211.h
+++ b/include/uapi/linux/nl80211.h
@@ -1502,6 +1502,8 @@ enum nl80211_commands {
 
 	NL80211_CMD_ASSOC_COMEBACK,
 
+	NL80211_CMD_STOP_BGRADAR_DETECT,
+
 	/* add new commands above here */
 
 	/* used to define NL80211_CMD_MAX below */
diff --git a/net/wireless/nl80211.c b/net/wireless/nl80211.c
index 7192cf8..6f3eb12 100644
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -9973,6 +9973,17 @@ unlock:
 	return err;
 }
 
+static int nl80211_stop_radar_detection(struct sk_buff *skb,
+						struct genl_info *info)
+{
+	struct net_device *dev = info->user_ptr[1];
+	struct wireless_dev *wdev = dev->ieee80211_ptr;
+
+	cfg80211_stop_background_radar_detection(wdev);
+
+	return 0;
+}
+
 static int nl80211_notify_radar_detection(struct sk_buff *skb,
 					  struct genl_info *info)
 {
@@ -16829,6 +16840,14 @@ static const struct genl_small_ops nl80211_small_ops[] = {
 		.flags = GENL_UNS_ADMIN_PERM,
 		.internal_flags = NL80211_FLAG_NEED_NETDEV_UP,
 	},
+	{
+		.cmd = NL80211_CMD_STOP_BGRADAR_DETECT,
+		.validate = GENL_DONT_VALIDATE_STRICT | GENL_DONT_VALIDATE_DUMP,
+		.doit = nl80211_stop_radar_detection,
+		.flags = GENL_UNS_ADMIN_PERM,
+		.internal_flags = NL80211_FLAG_NEED_NETDEV_UP |
+					 NL80211_FLAG_NO_WIPHY_MTX,
+	},
 };
 
 static struct genl_family nl80211_fam __genl_ro_after_init = {
@@ -18646,6 +18665,12 @@ nl80211_radar_notify(struct cfg80211_registered_device *rdev,
 			goto nla_put_failure;
 	}
 
+	if (rdev->background_radar_wdev &&
+	    cfg80211_chandef_identical(&rdev->background_radar_chandef, chandef)) {
+		if(nla_put_flag(msg, NL80211_ATTR_RADAR_BACKGROUND))
+			goto nla_put_failure;
+	}
+
 	if (nla_put_u32(msg, NL80211_ATTR_RADAR_EVENT, event))
 		goto nla_put_failure;
 
-- 
2.17.1

